cmd/storer2elastic/dataprocessor/dataProcessor.go

```diff
@@ -0,0 +1,514 @@
+	var shardHeader block.Header
+	err := dp.marshalizer.Unmarshal(&shardHeader, shardHeaderBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &shardHeader, nil
+}
+
+func (dp *dataProcessor) indexHeader(persisters *persistersHolder, dbInfo *databasereader.DatabaseInfo, hdr data.HeaderHandler, epoch uint32) error {
+	if metaBlock, ok := hdr.(*block.MetaBlock); ok {
+		if metaBlock.IsStartOfEpochBlock() {
+			dp.processValidatorsForEpoch(metaBlock.Epoch, metaBlock, persisters.miniBlocksPersister)
+		}
+	}
+
+	miniBlocksHashes := make([]string, 0)
+	shardIDs := dp.getShardIDs()
+	for _, shard := range shardIDs {
+		miniBlocksForShard := hdr.GetMiniBlockHeadersWithDst(shard)
+		for hash := range miniBlocksForShard {
+			miniBlocksHashes = append(miniBlocksHashes, hash)
+		}
+	}
+
+	body, txPool, err := dp.processBodyAndTransactionsPoolForHeader(persisters, miniBlocksHashes)
+	if err != nil {
+		return err
+	}
+
+	singersIndexes, err := dp.computeSignersIndexes(hdr)
+	if err != nil {
+		return err
+	}
+	notarizedHdrs := dp.computeNotarizedHeaders(hdr)
+	dp.elasticIndexer.SaveBlock(body, hdr, txPool, singersIndexes, notarizedHdrs)
+	dp.indexRoundInfo(singersIndexes, hdr)
+
+	log.Info("indexed header", "epoch", hdr.GetEpoch(), "shard", hdr.GetShardID(), "nonce", hdr.GetNonce())
+	return nil
+}
+
+func (dp *dataProcessor) indexRoundInfo(signersIndexes []uint64, hdr data.HeaderHandler) {
+	ri := indexer.RoundInfo{
+		Index:            hdr.GetRound(),
+		SignersIndexes:   signersIndexes,
+		BlockWasProposed: false,
+		ShardId:          hdr.GetShardID(),
+		Timestamp:        time.Duration(hdr.GetTimeStamp()),
+	}
+
+	dp.elasticIndexer.SaveRoundsInfos([]indexer.RoundInfo{ri})
+}
+
+func (dp *dataProcessor) getShardIDs() []uint32 {
+	shardIDs := make([]uint32, 0)
+	for shard := uint32(0); shard < dp.shardCoordinator.NumberOfShards(); shard++ {
+		shardIDs = append(shardIDs, shard)
+	}
+	shardIDs = append(shardIDs, core.MetachainShardId)
+
+	return shardIDs
+}
+
+func (dp *dataProcessor) processBodyAndTransactionsPoolForHeader(
+	persisters *persistersHolder,
+	mbHashes []string,
+) (*block.Body, map[string]data.TransactionHandler, error) {
+	txPool := make(map[string]data.TransactionHandler)
+	mbUnit := persisters.miniBlocksPersister
+
+	blockBody := &block.Body{}
+	for _, mbHash := range mbHashes {
+		mbBytes, err := mbUnit.Get([]byte(mbHash))
+		if err != nil {
+			log.Warn("miniblock not found in storage", "hash", []byte(mbHash))
+			continue
+		}
+
+		recoveredMiniBlock := &block.MiniBlock{}
+		err = dp.marshalizer.Unmarshal(recoveredMiniBlock, mbBytes)
+		if err != nil {
+			log.Warn("cannot unmarshal miniblock", "hash", mbHash, "error", err)
+			continue
+		}
+
+		dp.processTransactionsForMiniBlock(persisters, recoveredMiniBlock.TxHashes, txPool, recoveredMiniBlock.Type)
+		blockBody.MiniBlocks = append(blockBody.MiniBlocks, recoveredMiniBlock)
+	}
+
+	return blockBody, txPool, nil
+}
+
+func (dp *dataProcessor) processTransactionsForMiniBlock(
```

Comment:
This could have been a transaction processor

Reply:
for ratings I extracted a separate processor. I wouldn't do the same thing with transactions or something else because blocks, mini-blocks and transactions are tightly coupled and can remain in the same structure. Anyway, the structure is simpler after refactoring